collector/benchlib/src/benchmark.rs

```diff
@@ -102,19 +107,60 @@ impl BenchmarkGroup {
 }
 
 /// Adds a single benchmark to the benchmark group.
-/// ```ignore
-/// use benchlib::define_benchmark;
 ///
-/// define_benchmark!(group, my_bench, {
-///     || do_something()
-/// });
+/// The first argument should be the `BenchmarkGroup` into which will the benchmark be stored.
+/// The second argument should be the name of the benchmark (a string literal).
+/// The third argument should be an identifier of a function that will be benchmarked.
+///
+/// If you do not provide any other arguments, the benchmarked function should not receive any
+/// parameters (it can return something though).
+///
+/// If you provide a fourth argument, it should be an expression `init` that prepares
+/// input data for the benchmarked function. The function should then receive a single parameter of
+/// the same type as `init`.
+/// ```
+/// use benchlib::{benchmark_group, define_benchmark};
+///
+/// fn calc_no_param() -> u32 { 1 + 1 }
+///
+/// fn calc_param(a: u32) -> u32 {
+///     a + 1
+/// }
+///
+/// fn main() {
+///     benchmark_group(|group| {
+///         define_benchmark!(group, my_bench, calc_no_param);
+///         define_benchmark!(group, my_bench, calc_param, 1);
+///     });
+/// }
 /// ```
 #[macro_export]
 macro_rules! define_benchmark {
-    ($group:expr, $name:ident, $fun:expr) => {
-        let func = move || $fun;
-        $group.register(stringify!($name), func);
-    };
+    // Version without init
+    ($group:expr, $name:literal, $fun:ident) => {{
+        let constructor = || {
```

Comment:
I'm having trouble with all the nested closures. Are they all necessary? `constructor` is a zero-arg closure that returns a zero-arg closure. Feels like that could be collapsed into a single closure. And then `register` wraps the constructor in yet another closure.

Reply:
So I wanted to achieve this:
You have a constructor, which is a function that returns another function, and that returned function will be benchmarked. You should be able to call the constructor multiple times (e.g. because of wall-time/perf. counter measurement split or simply because we run more iterations). So that's why I need the first closure.

The second closure is needed to apply `black_box` on the return value of the benchmarked function. Before I could use `black_box` in the benchmark harness itself, but now that we pass a named function, I can't easily store an arbitrary function in a hashmap, because I couldn't name it's return type. So I apply `black_box` here directly on whatever the function returns, and then the created closure will just return `()`.

Reply:
In that case, please add more comments to the code explaining all this.